### ISSUE TITLE
Authentication for RPC

### ISSUE THREAD

**Issue by hatala91** [pytorch/pytorch#46904]
## 🚀 Feature
Consider a multi machine setup with an open network infrastructure. At the moment pyTorch RPC provides no way of securing the masters endpoint through authentication methods. This leads to following two vunrabilities:
- Attackers might query gradients/ weights from the master node to gain access to a model without owning the data.
- Anyone could disquise as a worker and send random gradients/ weights to the master to obscure the final model.

## Motivation
As a Software Engineer I want to provide solutions that ensure safety and security.

## Pitch
Make multi machine RPC training secure!

**Comment by pritamdamania87** [comment]
@hatala91 Thanks for providing feedback on the RPC framework! In terms of authentication and encryption, were you mostly looking for some sort of TLS based system for this or did you have something else in mind?

**Comment by hatala91** [comment]
Hi @pritamdamania87, thanks for the quick response. I had something like OAuth in mind for authentication (layer 7 in terms of the OSI layer model).

**Comment by wenjianhn** [comment]
Is there any plan to implement this feature? It would be really helpful for us.

**Comment by lw** [comment]
I don't think there's any plans to work on this. The RPC framework was designed for use cases inside fully user-controlled clusters, thus without strong security needs. I'd say that, currently, the best way to add some security is to implement a custom protocol in TensorPipe, which may encrypt the traffic based on some "secret" that's provided via env args or that's put in some filesystem location on each machine. However, implementing a TensorPipe protocol will be tricky, as it has to be done fully in C++, and there's no officially supported way to then add it to PyTorch, you'd need to do it hackily on your own.

**Comment by wenjianhn** [comment]
Thanks for the comment. We don't have time to do it yet.

We have a security scanner that sends mailformed RPC to pyTorch.
As a result, pyTorch often OOMs. It took us a week to find the root cause.

The reproducer:
```
import socket

HOST = 'localhost'

# Master port
PORT = 32540

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    s.connect((HOST, PORT))

    # enum class QueryType : uint8_t {
    QueryType_SET = b'\x00'
    s.sendall(QueryType_SET)

    # ~ 896.58 Gi
    s.sendall((962696183808).to_bytes(8, "little"))

    # Hung or reset
    s.sendall(b'0x00')
    data = s.recv(1024)
```
